### ISSUE TITLE
Better handling when loading coredump

### ISSUE THREAD

**Issue by leberus** [radare/radare2#9606]
cc @radare 

I've been thinking for a while that we should have better handling when loading coredumps.
Right now (I'm talking for ELF), the only thing we do is retrieve the registers and spit them to the output so the user has a chance to set it via "arw".
I think we could:

1) Retrieve the register values and set them automatically
2) Retrieve the maps and set them

The good thing about setting the maps is that the user could dig in the stack for example.

In order to do that, i think a core should be open with "-d".
Is there a preferred way to do this? (if not I can do it my way)

**Comment by radare** [comment]
Agree. Theres no prefered way. Just design it the best you can think of and we can discuss later on this :)

**Comment by XVilka** [comment]
Keep in mind that the same treatment may be needed for MDMP loading too.

**Comment by leberus** [comment]
You can assign this to me if you want

**Comment by leberus** [comment]
Ok, I have got something.
The approach I took is to split it in two parts:

<b>1)</b> $ r2 [corefile] - This will load the maps and the regstate found in the coredump. Only for analysis.
<b>2)</b> $ r2 -d [binary] [corefile] - This will start a new process and set the maps and the regstate to the ones found in the coreudmp. This will be for debugging.

So far, I got the first part working:

<pre>
$ /lab/r2/core> readelf -h core | grep Type
  Type:                              CORE (Core file)
$ /lab/r2/core> r2 core 
Setting up coredump: 17 maps have been found and created
Setting up coredump: Registers have been set
 -- Now with more better English!
[0x00000000]> om
17 fd: 3 +0x0003807c 0xffffffffff600000 - 0xffffffffff600fff -r-x 
16 fd: 3 +0x0003607c 0x7ffd7d93b000 - 0x7ffd7d93cfff -r-x 
15 fd: 3 +0x0003607c 0x7ffd7d938000 - 0x7ffd7d93afff -r-x 
14 fd: 3 +0x0001507c 0x7ffd7d880000 - 0x7ffd7d8a0fff -r-x 
13 fd: 3 +0x0001407c 0x7f1c7983e000 - 0x7f1c7983efff -r-x 
12 fd: 3 +0x0001307c 0x7f1c7983d000 - 0x7f1c7983dfff -r-x /lib64/ld-2.22.so
11 fd: 3 +0x0001207c 0x7f1c7983c000 - 0x7f1c7983cfff -r-x /lib64/ld-2.22.so
10 fd: 3 +0x0001007c 0x7f1c7983a000 - 0x7f1c7983bfff -r-x 
 9 fd: 3 +0x0000d07c 0x7f1c79819000 - 0x7f1c7981bfff -r-x 
 8 fd: 3 +0x0000d07c 0x7f1c7961c000 - 0x7f1c7963cfff -r-x /lib64/ld-2.22.so
 7 fd: 3 +0x0000907c 0x7f1c79618000 - 0x7f1c7961bfff -r-x 
 6 fd: 3 +0x0000707c 0x7f1c79616000 - 0x7f1c79617fff -r-x /lib64/libc-2.22.so
 5 fd: 3 +0x0000307c 0x7f1c79612000 - 0x7f1c79615fff -r-x /lib64/libc-2.22.so
 4 fd: 3 +0x0000307c 0x7f1c79279000 - 0x7f1c79411fff -r-x /lib64/libc-2.22.so
 3 fd: 3 +0x0000207c 0x00601000 - 0x00601fff -r-x /home/osalvador/lab/r2/core/arg
 2 fd: 3 +0x0000107c 0x00600000 - 0x00600fff -r-x /home/osalvador/lab/r2/core/arg
 1 fd: 3 +0x0000107c 0x00400000 - 0x00400fff -r-x /home/osalvador/lab/r2/core/arg
[0x00000000]> [0x00000000]> ar
rax = 0xfffffffffffffdfc
rbx = 0x7ffd7d89fc00
rcx = 0x7f1c79333040
rdx = 0x00000000
rsi = 0x7ffd7d89fbf0
rdi = 0x7ffd7d89fbf0
r8 = 0x7ffd7d89fd00
r9 = 0x7ffd7d89fb40
r10 = 0x00000008
r11 = 0x00000246
r12 = 0x7ffd7d89fc80
r13 = 0x7ffd7d89fed0
r14 = 0x00000000
r15 = 0x00000000
rip = 0x7f1c79333040
rbp = 0xffffffff
rflags = 0x00000246
rsp = 0x7ffd7d89fbe8
[0x00000000]> px 8 @ 0x7ffd7d89fbe8
- offset -       0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x7ffd7d89fbe8  f42e 3379 1c7f 0000                      ..3y....
[0x00000000]> pd 4 @ 0x004004f0
            0x004004f0      31ed           xor ebp, ebp
            0x004004f2      4989d1         mov r9, rdx
            0x004004f5      5e             pop rsi
            0x004004f6      4889e2         mov rdx, rsp
[0x00000000]>
</pre>

I have only tested it on x86_64, so the next step is to check that works on other archs.
After that, I will start part 2.

**Comment by radare** [comment]
Cool!

**Comment by radare** [comment]
can we close this now?

**Comment by leberus** [comment]
No please.
I would like to start with the part 2 of this, which should be loading the binary for debugging with the status of the coredump, just as gdb does with "gdb [binary] [coredump]".
I am planning to start soon.

**Comment by radare** [comment]
ping

**Comment by mcd1992** [comment]
Commenting so I can find this later and give implementation a shot after getting a little more familiar with the codebase.
